### ISSUE TITLE
Negative Sample Selection Issue

### ISSUE THREAD

**Issue by Mistsink** [DeepGraphLearning/ULTRA#26]
Hello, your work is extremely valuable. After reading your paper, I understand that in the link prediction task, the score for each node being a tail node is calculated based on the information propagated outward from the head node, correct? In your experimental setup, the GNN has 6 layers, so does this mean that you only consider nodes within the 6-hop neighborhood subgraph of the head node as candidate nodes (negative samples)? In this case, nodes outside the 6-hop subgraph would not be considered. If the correct tail node is outside the 6-hop neighborhood, this method is destined to miss the correct tail node. Is my understanding correct? I look forward to your response and greatly appreciate your feedback. 🙏

**Comment by migalkin** [comment]
Hi, thanks for the kind words!

>  the score for each node being a tail node is calculated based on the information propagated outward from the head node, correct?

Correct

> only consider nodes within the 6-hop neighborhood subgraph of the head node as candidate nodes (negative samples)?

By default, we only sample "hard negatives" - that is, if the current query triple to be predicted is `(h, r, t_1)` and there are other edges in the graph starting from the same head node with the same relation `(h, r, t_2), (h, r, t_3)`, we remove `t_2` and `t_3` from the sampling pool. You can check the negative sampling procedure [here](https://github.com/DeepGraphLearning/ULTRA/blob/561f2c79abaf51901702d1712554c35e3739bb8d/ultra/tasks.py#L42)

> If the correct tail node is outside the 6-hop neighborhood, this method is destined to miss the correct tail node.

Technically yes, but practically no - most dense graphs have a diameter <7, and in nearly all KGs true tails are <=4 hops away from the head node. We tried training a model with 10 layers but it didn't yield any significant improvement.

**Comment by Mistsink** [comment]
Thank you very much for your prompt reply, haha. I hope this issue won't disrupt your daily work.

I am very aware of the strict negative sampling strategy in your code, which is very reasonable and accurate for our link prediction task. I have no doubts about this aspect.

However, in the comparison of experimental results, if we only consider all negative samples within 6 hops of the head node (ignoring nodes beyond 6 hops), the predicted results would actually be equivalent to your experimental method and thus comparable. Is my understanding correct?

**Comment by migalkin** [comment]
You might be mixing negative sampling and final predictions/evaluations. We train on `k` negative samples, but validation/testing is done on the whole set of entities in the graph.

Right now negative samples _can_ be more than 6 hops away if the graph diameter is large. I don't think that restricting the pool of candidate negatives to <6 hops would play a major role in training as long as the negatives are "hard negatives". However, if you restrict the set of entities _at_ _evaluation_, then the results won't be comparable, because metrics will be computed on a different number of nodes. For example, 0.4 MRR on 15k nodes on FB15k237 and 0.4 MRR on 5k nodes on a subset of FB15k237 are different values (and 0.4 MRR on 5k in this case is practically worse)

**Comment by Mistsink** [comment]
What if we don't consider MR and MRR, and only compare the results of hit@1, 3, 5, and 10? If the correct tail node is within the 6-hop neighborhood of the head, I will calculate the scores of all nodes within 6 hops (of course, filtering out tail nodes that also satisfy the relation, such as the t1 in h→r→t1) and compute the hit result for the correct tail node. If the correct tail node is outside the 6-hop neighborhood, its hit result will be 0. Do you think the hit scores obtained in this way are comparable?

**Comment by migalkin** [comment]
Still not comparable - your model might be bad enough to assign exactly the same score for the true tail and, for example, 100 other nodes where 50 of them happen to be outside the 6-hop neighborhood. If you artificially cut those 50 scores setting them to 0, then you artificially bump the rank of the prediction. 

Please don't rig standard evaluation protocols :)

**Comment by Mistsink** [comment]
I might not have explained my point clearly. Let me give an example: during evaluation, for the first query h1,r1,t1, if t1 is within the 6-hop neighborhood of h1, I will calculate the scores of all candidate nodes within 6 hops, setting the scores of all nodes outside the 6-hop neighborhood to 0, because under your method's assumption, they are destined not to receive information from h1. Based on this assumption, I directly filter out the nodes outside the 6-hop neighborhood. For the second query h2,r2,t2, if t2 is outside the 6-hop neighborhood of h2, based on the above assumption, I directly consider t2 to be unpredicted, so its hit score will also be set to 0.
Does this make my approach clearer to you?

**Comment by Mistsink** [comment]
Or do you mean that I can only randomly select 50 candidate samples and compare hit results with methods like GraIL?

**Comment by migalkin** [comment]
> Or do you mean that I can only randomly select 50 candidate samples and compare hit results with methods like GraIL?

No, this is the worst setup ever, don't use this - it overestimates the true performance.

> Does this make my approach clearer to you?

Yes it does, and I am still convinced the numbers will be different. Actually, let's run a simple experiment - because we can't argue with the numbers :) 

A standard performance of ULTRA 4g on WN18RR test (which I just ran) is:
```
01:46:42   Evaluate on test
01:52:26   mr: 837.171
01:52:26   mrr: 0.483882
01:52:26   hits@1: 0.414805
01:52:26   hits@3: 0.519783
01:52:26   hits@10: 0.617422
```

Now let's mask out all entities further than 6 hops from queried head/tail nodes, for that, in between lines 143 and 144
https://github.com/DeepGraphLearning/ULTRA/blob/561f2c79abaf51901702d1712554c35e3739bb8d/script/run.py#L143-L144

let's insert the following snippet:
```python
for i in range(len(batch)):
    t_subset, _, _, _ = pyg.utils.k_hop_subgraph(t_batch[i, 0, 0].view(1,), 6, test_data.edge_index, num_nodes=test_data.num_nodes)
    new_t_mask = torch.zeros(t_mask[i].shape, dtype=torch.bool, device=device)
    new_t_mask[t_subset] = True
    # mask all nodes outside the 6-hop neighborhood 
    t_mask[i][~new_t_mask] = False
    h_subset, _, _, _ = pyg.utils.k_hop_subgraph(h_batch[i, 0, 0].view(1,), 6, test_data.edge_index, num_nodes=test_data.num_nodes)
    new_h_mask = torch.zeros(t_mask[i].shape, dtype=torch.bool, device=device)
    new_h_mask[h_subset] = True
    # mask all nodes outside the 6-hop neighborhood 
    h_mask[i][~new_h_mask] = False
```

Now let's run eval on WN18RR once again and we see very different and much better performance:
```
01:40:00   Evaluate on test
01:46:02   mr: 204.884
01:46:02   mrr: 0.576971
01:46:02   hits@1: 0.508137
01:46:02   hits@3: 0.614391
01:46:02   hits@10: 0.703414
```

I think we reached the bottom line here :)

**Comment by Mistsink** [comment]
Oh, thank you so much. I really appreciate that you went through the trouble of running the experiment to explain it to me. I've never met such a patient person online before, haha.

Additionally, does this experiment indicate that other negative samples outside the 6-hop neighborhood might also have scores higher than the correct tail node? Therefore, after masking, the hit score would increase. In other words, although theoretically a 6-layer GNN would prevent nodes outside the 6-hop neighborhood from obtaining information about the head, it is still possible for their scores to be higher than those within the 6-hop neighborhood.

**Comment by migalkin** [comment]
Don't forget that the final score for each node is obtained _not_ just from the final node state after message passing, but after an MLP that concatenates a query vector with the final node state:

https://github.com/DeepGraphLearning/ULTRA/blob/561f2c79abaf51901702d1712554c35e3739bb8d/ultra/models.py#L170

That is, even if a state of some node is 0 after message passing, and MLP over `[0, query]` might return some plausible result after training (eg, if training included predicting an edge between disconnected nodes). 
So yes, it is still possible.

**Comment by Mistsink** [comment]
Thank you very much! Your explanation has given me a deeper understanding of your method, and it has also broadened my perspective on deep learning. It's no exaggeration to say that this issue will undoubtedly have a subtle yet profound impact on my future AI studies (if I continue with AI, haha). Once again, thank you so much, and I hope to have the opportunity to meet and thank you in person.
